umapi_client/functional.py

```diff
@@ -308,6 +320,48 @@ def delete_account(self):
         self.append(removeFromDomain={})
         return None
 
+    def maybe_split_groups(self, max_groups):
+        """
+        Check if group lists in add/remove directives should be split and split them if needed
+        :param max_groups: Max group list size
+        :return: True if at least one command was split, False if none were split
+        """
+        new_commands = []
+        # return True if we split at least once
+        maybe_split = False
+        for command in self.commands:
+            step_key, step_args = next(six.iteritems(command))
+            if step_key not in [StepKeys.add.name, StepKeys.remove.name, StepKeys.addRoles.name]:
+                continue
+            split_commands = self._split_groups(step_key, command, max_groups)
+            if len(split_commands) > 1:
+                maybe_split = True
+            new_commands += split_commands
+        self.commands = new_commands
+        return maybe_split
+
+    def _split_groups(self, step_key, command, max_groups):
+        """
+        Split long group lists of individual command
+        Creates additional commands for each split and recursively splits new groups until no commands exceed group
+        limit.  Assumes that at least one split should occur.
+        :param step_key: Step key (add, remove, etc)
+        :param command: Single command containing one or more group lists to add, remove, etc
+        :param max_groups: Max group list size
+        :return: List of at least two commands that resulted from split
+        """
+        new_commands = [command]
+        while True:
+            new_command = {step_key: {}}
+            for group_type, groups in six.iteritems(command[step_key]):
+                if len(groups) > max_groups:
+                    command[step_key][group_type], new_command[step_key][group_type] = \
+                        groups[0:max_groups], groups[max_groups:]
+            if new_command[step_key]:
+                new_commands += self._split_groups(step_key, new_command, max_groups)
```

Comment:
There's no need to recurse; this is already in a `while True` loop.  Instead just append the command to `new_commands` and set `command` to new_command and let the loop go again.  The bug in my prior code was not setting `command` to `new_command`.  And then you can just fold this into the upper function, setting the `maybe_split` to true when you split.